### PR TITLE
feat: MLIBZ-2558 Change how sync queue stores request information

### PR DESCRIPTION
**Pull request by yuliya-guseva**
#### Description
Right now, requests for the sync queue are only being saved as a SAVE or DELETE. It would be better to expose out what type of request we are pursuing (POST/PUT/DELETE) in the queue itself, in order to be able to replace temporary IDs with permanent IDs easier.

#### Changes
Changed sync items types from `SAVE or DELETE` to `POST/PUT/DELETE`. `SAVE` type was saved for backward compatibility with previous implementation.

#### Tests
Instrumented

### CODE REVIEW

File: java-api-core/src/com/kinvey/java/network/NetworkManager.java [hunk 1]
```diff
@@ -468,6 +462,18 @@ public String onUnsavedReferenceFound(String collection, GenericJson object) {
         return save;
     }
 
+    public boolean isTempId(T item) {
+        boolean isTempId = true;
+        if (item.get(Constants._ID) != null) {
```
Comment (vinaygahlawat) [inline java-api-core/src/com/kinvey/java/network/NetworkManager.java]: Why is the default value here `true`?  Would it be better if it were `false`, so we only return `true` if we are certain that the format of the `_id` is the temporary format?

File: java-api-core/src/com/kinvey/java/network/NetworkManager.java [hunk 2]
```diff
@@ -468,6 +462,18 @@ public String onUnsavedReferenceFound(String collection, GenericJson object) {
...
+            try {
+                isTempId = item.get(Constants._ID).toString().matches("\\w{8}-\\w{4}-\\w{4}-\\w{4}-\\w{12}");
```
Comment (vinaygahlawat) [inline java-api-core/src/com/kinvey/java/network/NetworkManager.java]: Does this also have to check that the `_id` starts with `temp_`?

File: java-api-core/src/com/kinvey/java/store/requests/data/PushRequest.java [hunk 3]
```diff
@@ -81,7 +86,16 @@ public Void execute() throws IOException {
                     }
                 }
                 try {
-                    syncManager.executeRequest(client, syncRequest);
+                    if (httpVerb == SyncRequest.HttpVerb.POST) {
+                        String tempID = syncRequest.getEntityID().id;
+                        GenericJson result = syncManager.executeRequest(client, syncRequest);
+                        T temp = cache.get(tempID);
+                        temp.set("_id", result.get("_id"));
```
Comment (vinaygahlawat) [inline java-api-core/src/com/kinvey/java/store/requests/data/PushRequest.java]: Would be good to use a constant string value here for `_id`, such as `Constants._ID`.

File: java-api-core/src/com/kinvey/java/sync/SyncManager.java [hunk 4]
```diff
@@ -159,35 +161,50 @@ public void enqueueRequest(SyncRequest request) {
         requestCache.save(request);
     }
 
-    public <T extends GenericJson> void enqueueRequest(String collectionName, NetworkManager<T> networkManager, RequestMethod method, String id) throws IOException {
+    public <T extends GenericJson> void enqueueRequest(String collectionName, NetworkManager<T> networkManager, SyncItem.HttpVerb httpMethod, String id) throws IOException {
         ICache<SyncItem> requestCache = cacheManager.getCache(SYNC_ITEM_TABLE_NAME, SyncItem.class, Long.MAX_VALUE);
```
Comment (vinaygahlawat) [inline java-api-core/src/com/kinvey/java/sync/SyncManager.java]: Are some of these API changes potentially breaking changes to the public API?

File: java-api-core/src/com/kinvey/java/network/NetworkManager.java (hunk 2 above)
Comment (yuliya-guseva) [reply java-api-core/src/com/kinvey/java/network/NetworkManager.java]: Yes, this has to check that the `_id` starts with ` temp_`, but it was the feature from another branch which was merged in the current branch today, so currently the SDK already checks ` temp_`. If a user had items which weren't synced the items will be pushed using `PUT`, but all new items will be pushed with `POST` request. The PR has instrumented test with such case.

File: java-api-core/src/com/kinvey/java/store/requests/data/PushRequest.java (hunk 3 above)
Comment (yuliya-guseva) [reply java-api-core/src/com/kinvey/java/store/requests/data/PushRequest.java]: I agree. It was updated.

File: java-api-core/src/com/kinvey/java/network/NetworkManager.java (hunk 1 above)
Comment (yuliya-guseva) [reply java-api-core/src/com/kinvey/java/network/NetworkManager.java]: You are right. And if the SDK gets `true` from this method then the POST request will be made and if it was not temp id then we will have duplicate items at the backend with different ids. Updated.

File: java-api-core/src/com/kinvey/java/sync/SyncManager.java (hunk 4 above)
Comment (yuliya-guseva) [reply java-api-core/src/com/kinvey/java/sync/SyncManager.java]: We don't have any notes about this method or this class at the devcenter. It's class for internal SDK using. I'm not sure that we have to add this changes to `breaking changes` for release notes. These changes don't affect classes and methods which exist in devcenter notes.

**Pull request merged**